configure.ac

```diff
@@ -44,59 +58,122 @@ AC_PROG_CPP
+GX_FC_QUAD_PRECISION()
 
-# Require netCDF.
-#AC_CHECK_FUNC([nf_open], [], [AC_MSG_ERROR([NetCDF Fortran library required to build FMS])])
+# Check if Fortran compiler supports reading namelists from internal files
+GX_FC_INTERNAL_FILE_NML()
 
-# Do we want to try and set Fortran flags for the user?
-if test $enable_fortran_flag_setting = "yes"; then
-   # Make sure the compiler is seeing this as free-formatted, not
-   # fixed-formatted, fortran code.
-   AC_FC_FREEFORM()
+if test $with_openmp = yes && test $enable_setting_flags = yes; then
+  # Check if the compiler needs special OpenMP flags
+  AC_OPENMP()
+fi
+AC_LANG_POP(Fortran)
 
-   # Check that long lines of Fortran code can be handled. This will add
-   # necessary fortran flags.
-   AC_FC_LINE_LENGTH([unlimited])
+# We passed all the tests.  Set the required defines.
+AC_DEFINE([use_netCDF], [1], [This is required for the library to build])
+if test $with_mpi = yes; then
+  AC_DEFINE([use_libMPI], [1], [This is required for the library to build])
 fi
 
-# These defines are required for the build.
-AC_DEFINE([use_netCDF], [1])
-AC_DEFINE([use_libMPI], [1])
+# Set any required compile flags.  This will not be done if the user wants to
+# define all their own flags.
+if test $enable_setting_flags = yes; then
+  # Make sure the compiler is seeing this as free-formatted, not
+  # fixed-formatted, fortran code.
+  AC_FC_FREEFORM()
+
+  # Check that long lines of Fortran code can be handled. This will add
+  # necessary fortran flags.
+  AC_FC_LINE_LENGTH([unlimited])
+
+  # Will we build with default 64-bit reals in Fortran, or do mixed mode?
+  if test $enable_mixed_mode = yes; then
+    GX_FC_DEFAULT_REAL_KIND4_FLAG([dnl
+      FCFLAGS="$FCFLAGS $FC_DEFAULT_REAL_KIND8_FLAG"])
+    AC_DEFINE([OVERLOAD_R4], [1], [Set to overload the R4 Fortran routines])
+    AC_DEFINE([OVERLOAD_R8], [1], [Set to overload the R8 Fortran routines])
+  else
+    GX_FC_DEFAULT_REAL_KIND8_FLAG([dnl
+      FCFLAGS="$FCFLAGS $FC_DEFAULT_REAL_KIND8_FLAG"])
+  fi
+
+  # Add Cray Pointer support flag
+  if test ! -z "$FC_CRAY_POINTER_FLAG"; then
+    FCFLAGS="$FCFLAGS $FC_CRAY_POINTER_FLAG"
+  fi
+
+  # Add OpenMP flags
+  if test ! -z "$OPENMP_CFLAGS"; then
+    CFLAGS="$CFLAGS $OPENMP_CFLAGS"
+    CPPFLAGS="$CPPFLAGS $OPENMP_CFLAGS"
+  fi
+  if test ! -z "$OPENMP_FCFLAGS"; then
+    FCFLAGS="$FCFLAGS $OPENMP_FCFLAGS"
+  fi
+fi
 
+# Settings based on the system and hardware
 # Define __APPLE__ macro if on a Apple/Mac OS X
 AS_IF([test "x$target_vendor" = "xapple"], [AC_DEFINE([__APPLE__], [ ])])
-
-# Define an AM_CONDITIONAL to determine if you are on a CRAY
-AM_CONDITIONAL([CRAY], [test `env | grep CRAY | wc -l` -gt 0])
-
 # Define an AM_CONDITIONAL to determine if you are on an Apple/Mac OS X
 AM_CONDITIONAL([APPLE], [test "x$target_vendor" = "xapple"])
+# Define an AM_CONDITIONAL to determine if you are on a CRAY
+AM_CONDITIONAL([CRAY], [test `env | grep CRAY | wc -l` -gt 0])
```

Comment:
I cannot overstate what a terrible idea this is. Do NOT check what kind of system you are on, and write code based on that. Instead check for the features you need, and program based on that.

What is it about a Cray that makes you think you need to know when you are compiling on the cray? What specifically?

All Crays are not the same. And other systems may also exhibit the same problem you are coding around. So figure out how to test whether the system supports what you need, and ignore the type of system. Ask for help if you need it to create the test.

Reply:
We are looking for was to better implement these checks.  Some of it requires understanding the requirement(s) that cause things to not build or work properly on these specific platforms/systems.  For the Apple case, we know it is a missing C call that is different on the Darwin/Mac OS systems.  It may affect other BSD-style OS's, but we do not yet know, but at the time the problem was discovered we only found it on Mac OS systems.

The Cray case is more difficult to discover via other tests.  It has to do with the use of shared libraries on systems that use the Cray Programming Environment (PrgEnv).  We are still investigating ways to reliably capture the problem via configure tests.  Not all Cray systems use the Cray PrgEnv.  Those systems should not cause this conditional to be set.  Once we have a reliable way to capture this, this will be updated.

Reply:
OK, is this only a problem for shared library builds?

There is already a --disable-shared option automatically built into configure.

Suggest you change the name of this from CRAY to something like CRAY_PRGENV_SHARED (or anything more specific). Otherwise, another programmer is going to come along, and see "CRAY" and start writing code based on that. Which is what you don't want. The more specific you make your macro names, the less temptation to misuse them later.

Also, from what you say, there will be some Cray systems that will not set CRAY because they don't have this issue. So that's a confusing macro name.